### PR TITLE
Adding host discovery

### PR DESCRIPTION
**Pull request by Mzack9999**
## Description
This PR implements various hosts discovery techniques:
- [x] icmp echo request
- [x] icmp timestamp request
- [x] icmp address mask request
- [x] ARP
- [x] IPv6 Neighbor Discovery (prototype via ICMPv6 - tracked at https://github.com/projectdiscovery/naabu/issues/351)
- [x] Tcp Syn Probes
- [x] Tcp Ack Probes
- [x] ~~IP Protocols (tracked at https://github.com/projectdiscovery/naabu/issues/353)~~
- [x] ~~Udp Probes (Depends on https://github.com/projectdiscovery/naabu/issues/305)~~
- [x] ~~IP reverse PTR lookup (tracked at https://github.com/projectdiscovery/naabu/issues/352)~~

## Notes
* Define if host discovery should just list discovered hosts and/or use them for subsequent ports scanning?

### CODE REVIEW

**Comment by ehsandeep** [comment]
> Define if host discovery should just list discovered hosts and/or use them for subsequent ports scanning?

- By default, we can keep host discovery and port scanning independent and optional.
- If **host discovery** (`sn`) flag is used without specifying any available methods, we can automatically use best known methods for discovery, otherwise specified method will be used.
- If **host discovery** (`sn`) flag is used with`p` flag, port scan will be performed on discovered hosts.
- for the cases when `sn` flag is not used, the current behavior will remain same, i.e performing syn/connect-based port scan.

@Mzack9999 ^

**Comment by Mzack9999** [comment]
> * If **host discovery** (`sn`) flag is used with the `p` flag, port scan will be performed on discovered hosts.

This can't be implemented as is since the `-p` CLI argument is not a boolean flag but a string one and requires a value different from `""`, otherwise the CLI help would be shown:
```console
$ echo 192.168.1.1 | go run . -p
flag needs an argument: -p
```

**Review by ehsandeep — DISMISSED** [review]
- [x] Validation check / error for user permission as host discovery requires sudo

File: v2/pkg/runner/runner.go [hunk 1]
```diff
@@ -523,12 +608,13 @@ func (r *Runner) SetInterface(interfaceName string) error {
 	return nil
 }
 
-func (r *Runner) handleOutput() {
+func (r *Runner) handleOutput(scanResults *result.Result) {
```
Comment (Ice3man543) [inline v2/pkg/runner/runner.go:611]: maybe we should split this function into multiple individual function as it's a bit hard to read

**Review by Ice3man543 — APPROVED** [review]

**Pull request merged**